### ISSUE TITLE
Snapshot testing

### ISSUE THREAD

**Issue by czosel** [glayzzle/php-parser#152]
Currently, a typical test looks roughly like this:
```js
  it("test coalesce operator", function() {
    var ast = parser.parseEval("$var = $a ?? true;");
    ast.children[0].right.kind.should.be.exactly("bin");
    ast.children[0].right.type.should.be.exactly("??");
    ast.children[0].right.left.kind.should.be.exactly("variable");
    ast.children[0].right.right.kind.should.be.exactly("boolean");
});
```
After working with snapshot-based testing for a while in prettier/plugin-php, I've really come to appreciate not having to write my own assertions anymore.

If we were to introduce snapshot-based testing here, the test case above would essentialy boil down to just one line, which could even be stored in a `.php` file:
```php
$var = $a ?? true;
```
The testing framework would then generate the snapshot and save it in a file:
```json
{
  "kind": "program",
  "children": [
    {
      "kind": "assign",
      "operator": "=",
      "left": {
        "kind": "variable",
        "name": "var",
        "byref": false,
        "curly": false
      },
      "right": {
        "kind": "bin",
        "type": "??",
        "left": {
          "kind": "variable",
          "name": "a",
          "byref": false,
          "curly": false
        },
        "right": {
          "kind": "boolean",
          "value": true,
          "raw": "true"
        }
      }
    }
  ],
  "errors": []
}
```
Like this, we'd see the exact AST changes that every PR introduces to any of the existing test cases.

It seems that [Esprima is using a similar approach](https://github.com/jquery/esprima/tree/master/test/fixtures/ES2016/exponent).

If you're interested, I could try preparing a little proof of concept - let me know what you think! 😄

**Comment by ichiriac** [comment]
Hi @czosel 

I like the idea of make tests based on snapshots, I miss them here over the AST and it could help to increase easily my code coverage.

By the way, I've actually done some work in the same direction based on tokens :
- here the testing framework : https://github.com/glayzzle/php-parser-dev
- here snapshoots of various frameworks : https://github.com/glayzzle/php-parser-dev-tokens

Sadly I have not enough time to work on the project but I would be very glad if you could help me with that.

**Comment by czosel** [comment]
Hi @ichiriac, thanks for your quick response - I didn't know about the two testing repos yet. I guess that integrating snapshot testing in this repo and basically porting the existing test cases would already be a great first step.

I hope I'll find the time soon to give this a shot!

**Comment by DaGhostman** [comment]
Would love to be of help, I'll try my best to find some spare time and help with whatever I can :)

**Comment by b4dnewz** [comment]
hi everybody, I looked to this snapshot testing and I think is a wonderful idea!
I've a suggestion regarding the **fixture.js** file, right now if the parsing produces an error it still write the file, I guess is like so for testing the errors too, am i right?

otherwise my proposal is to write the .json output only if the parsing succeed so you don't have to manually remove it after you fix the php file template issue.

```js
try {
  // parse the file
  ast = parser.parseCode(buffer, filename, options);
  ast = JSON.stringify(ast, null, 2);

  try {
    original = fs.readFileSync(filename + ".json").toString();
  } catch (e) {
    console.log("Generating snapshot: " + filename);
    // GENERATE THE OUTPUT (IF FILE DOES NOT EXISTS)
    original = ast;
    fs.writeFileSync(filename + ".json", original);
  }

  original.should.be.exactly(ast, "Fail " + filename);
} catch (e) {
  console.log("Failed file parsing for:", filename);
  console.error(e);
}
```
otherwise if is needed for testing errors it's good as it is.

also I see is a very huge work to do and I would like some tips from @ichiriac about where to start and how to deal these tests when there are more than one ast parsing for describe method like [this one](https://github.com/glayzzle/php-parser/blob/master/test/commentTests.js#L92-L174) without loosing the readability of the tests.

maybe a recursive fixture.js function that look for subfolders and creates describe blocks.
by the way, this snapshot testing is wery cool! 😎

**Comment by czosel** [comment]
@b4dnewz I'd suggest using an existing testing framework to do this, for example [jest](https://facebook.github.io/jest/docs/en/snapshot-testing.html). I wanted to prepare a little proof of concept, but didn't find the time yet :disappointed:

**Comment by ichiriac** [comment]
Hi,

I've started a new branch here, and switched all tests over jest : https://github.com/glayzzle/php-parser/tree/v3.1

You can make your PR there :)

**Comment by ichiriac** [comment]
thanks @b4dnewz, I've merged your [PR #157](https://github.com/glayzzle/php-parser/pull/157), all tests are under Jest and the code coverage is great, you've done an amazing work !

Many thanks dude :+1:

**Comment by ichiriac** [comment]
This fix is now relased under 3.0.0-alpha.3
